airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/internals/mssql/MssqlDebeziumStateUtil.java

```diff
@@ -0,0 +1,157 @@
+import java.time.Duration;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Stream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MssqlDebeziumStateUtil {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(MssqlDebeziumStateUtil.class);
+
+  public JsonNode generateOffsetJsonNode(
+                                         final JdbcDatabase database) {
+    final AirbyteFileOffsetBackingStore offsetManager = AirbyteFileOffsetBackingStore.initializeState(
+        constructLsnSnapshotState(database, database.getSourceConfig().get(JdbcUtils.DATABASE_KEY).asText()),
+        Optional.empty());
+    final Map<String, Object> state = new HashMap<>();
+    state.put(MssqlCdcStateConstants.MSSQL_CDC_OFFSET, offsetManager.read());
+    return Jsons.jsonNode(state);
+  }
+
+  public JsonNode constructInitialDebeziumState(final Properties properties,
+                                                final ConfiguredAirbyteCatalog catalog,
+                                                final JdbcDatabase database) {
+    final JsonNode highWaterMark = constructLsnSnapshotState(database, database.getSourceConfig().get(JdbcUtils.DATABASE_KEY).asText());
+    final AirbyteFileOffsetBackingStore emptyOffsetManager = AirbyteFileOffsetBackingStore.initializeState(null,
+        Optional.empty());
+    final AirbyteSchemaHistoryStorage schemaHistoryStorage =
+        AirbyteSchemaHistoryStorage.initializeDBHistory(new SchemaHistory<>(Optional.empty(), false), false);
+    final LinkedBlockingQueue<ChangeEvent<String, String>> queue = new LinkedBlockingQueue<>();
+    final Instant engineStartTime = Instant.now();
+    try {
+      final DebeziumRecordPublisher publisher = new DebeziumRecordPublisher(properties,
+          database.getSourceConfig(),
+          catalog,
+          emptyOffsetManager,
+          Optional.of(schemaHistoryStorage),
+          DebeziumPropertiesManager.DebeziumConnectorType.RELATIONALDB);
+      publisher.start(queue);
+      while (!publisher.hasClosed()) {
+        final ChangeEvent<String, String> event = queue.poll(10, TimeUnit.SECONDS);
+        if (event != null) {
+          publisher.close();
+          break;
+        }
+        if (Duration.between(engineStartTime, Instant.now()).compareTo(Duration.ofMinutes(3)) > 0) {
+          LOGGER.error("No record is returned even after {} seconds of waiting, closing the engine", 180);
+          publisher.close();
+          throw new RuntimeException(
+              "Building schema history has timed out. Please consider increasing the debezium wait time in advanced options.");
+        }
+      }
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
```

Comment:
Do we want to consider hoisting this into the CDK? This pattern appears in mysql also.

Reply:
This is in cdk, though not unified yet.